ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blobs/BlockBlobSidecarsTracker.java

```diff
@@ -0,0 +1,172 @@
+  }
+
+  public NavigableMap<UInt64, BlobSidecar> getBlobSidecars() {
+    return blobSidecars;
+  }
+
+  public SafeFuture<Void> getCompletionFuture() {
+    return blobSidecarsComplete;
+  }
+
+  public Optional<BeaconBlockBodyDeneb> getBlockBody() {
+    return blockBody.get();
+  }
+
+  public boolean containsBlobSidecar(final BlobIdentifier blobIdentifier) {
+    return Optional.ofNullable(blobSidecars.get(blobIdentifier.getIndex()))
+        .map(blobSidecar -> blobSidecar.getBlockRoot().equals(blobIdentifier.getBlockRoot()))
+        .orElse(false);
+  }
+
+  public Stream<BlobIdentifier> getMissingBlobSidecars() {
+    final Optional<BeaconBlockBodyDeneb> body = blockBody.get();
+    if (body.isPresent()) {
+      return UInt64.range(UInt64.ZERO, UInt64.valueOf(body.get().getBlobKzgCommitments().size()))
+          .filter(blobIndex -> !blobSidecars.containsKey(blobIndex))
+          .map(blobIndex -> new BlobIdentifier(slotAndBlockRoot.getBlockRoot(), blobIndex));
+    }
+
+    if (blobSidecars.isEmpty()) {
+      return Stream.of();
+    }
+
+    // We may return maxBlobsPerBlock because we don't know the block
+    return UInt64.range(UInt64.ZERO, maxBlobsPerBlock)
+        .filter(blobIndex -> !blobSidecars.containsKey(blobIndex))
+        .map(blobIndex -> new BlobIdentifier(slotAndBlockRoot.getBlockRoot(), blobIndex));
+  }
+
+  public Stream<BlobIdentifier> getUnusedBlobSidecarsForBlock() {
+    final Optional<BeaconBlockBodyDeneb> body = blockBody.get();
+    checkState(body.isPresent(), "Block must me known to call this method");
+
+    final UInt64 firstUnusedIndex = maxBlobsPerBlock.min(body.get().getBlobKzgCommitments().size());
+    return UInt64.range(firstUnusedIndex, maxBlobsPerBlock)
+        .map(blobIndex -> new BlobIdentifier(slotAndBlockRoot.getBlockRoot(), blobIndex));
+  }
+
+  public boolean add(final BlobSidecar blobSidecar) {
+    checkArgument(
+        blobSidecar.getBlockRoot().equals(slotAndBlockRoot.getBlockRoot()),
+        "Wrong blobSidecar block root");
+
+    if (blobSidecarsComplete.isDone()) {
+      // already completed
+      return false;
+    }
+
+    boolean addedNew = blobSidecars.put(blobSidecar.getIndex(), blobSidecar) == null;
+
+    if (addedNew) {
+      checkCompletion();
+    } else {
+      LOG.warn(
+          "Multiple BlobSidecars with index {} for {} detected.",
+          slotAndBlockRoot.toLogString(),
+          blobSidecar.getIndex());
+    }
+
+    return addedNew;
+  }
+
+  public int blobSidecarsCount() {
+    return blobSidecars.size();
+  }
+
+  public boolean setBlock(final SignedBeaconBlock block) {
+    checkArgument(block.getSlotAndBlockRoot().equals(slotAndBlockRoot), "Wrong block");
+    final Optional<BeaconBlockBodyDeneb> oldBlock =
+        blockBody.getAndSet(
+            Optional.of(BeaconBlockBodyDeneb.required(block.getMessage().getBody())));
+    if (oldBlock.isPresent()) {
+      LOG.debug("block was already set!");
```

Comment:
It's useless without any extra info (block info or whatever), and we could hit it very often. I'd remove it

Reply:
going to remove it for now, will put it in the debugging PR to show timings